### PR TITLE
Pass current run number to payu sync command.

### LINKED ISSUE #732: Automated payu sync command should pass along current run number
Currently payu sync is invoked here:

https://github.com/payu-org/payu/blob/1a083723d112bb6e748a08f510f84a9f0a072a89/payu/experiment.py#L998-L1001

For reference, collate passes along the current run number: 

https://github.com/payu-org/payu/blob/1a083723d112bb6e748a08f510f84a9f0a072a89/payu/experiment.py#L947-L951

I think the payu sync job currently syncs all outputs which could be an issue for jobs run in series (e.g. if a later payu run archives before an earlier sync job runs).

I think if the current run number is provided, it shouldn't sync any outputs from runs greater than current run number.

### PR DESCRIPTION
**Pull request by Qian-HuiChen**
This PR pass the current run number to `payu sync` command. In this case, restart and output greater than the current run will not be sync.

Close #732.

### CODE REVIEW

**Comment by Qian-HuiChen** [comment]
I’m not sure what is the best way to handle the payu_jobs directory gracefully.
At the moment, the entire payu_jobs directory is synced without filtering out jobs greater than the current run. 
However, if I do something like:
```
    def add_payu_jobs_to_sync(self):
        """Add paths of payu jobs in archive to sync. The last job is
        protected"""
        log_path = os.path.join(self.expt.archive_path, "payu_jobs")
        all_jobs = sorted(os.listdir(log_path)) if os.path.isdir(log_path) else []
        jobs = filter_previous_runs(all_jobs, prefix=None)
        jobs = [os.path.join(log_path, job) 
                for job in jobs]
        if len(jobs) > 0:
            last_job = jobs.pop()
            if not self.ignore_last:
                # Protect the last job
                self.source_paths.append(SourcePath(path=last_job,
                                                    protected=True))
        self.source_paths.extend([SourcePath(path, is_log_file=True) for path in jobs])
```
then the jobs get synced as `sync_path/0`, `sync_path/1`, etc., instead of `sync_path/payu_jobs/0`.

I can think of two ways:
1. changing the destination path for `payu_jobs`
2. sync the entire `payu_jobs` dir and set `--exclude` flag to exclude jobs greater than current run.

The tests in test_sync.py are not independent for now. I will improve them after #738 is merged.

**Comment by aidanheerdegen** [comment]
> I’m not sure what is the best way to handle the payu_jobs directory gracefully. At the moment, the entire payu_jobs directory is synced without filtering out jobs greater than the current run. However, if I do something like:

Do we need to handle `payu_jobs` in this way? Isn't it ok to just sync over the current state of `payu_jobs`?

File: payu/sync.py [hunk 1]
```diff
@@ -42,6 +42,25 @@ def __init__(self, path, protected=False, is_log_file=False):
         self.path = path
         self.is_log_file = is_log_file
 
+def filter_previous_runs(all_dir, prefix=None):
+    """Given a list of directories of all runs (e.g., ['output001', 'output002']), 
+    filter to only include dir that is less than or equal to current run."""
```
Comment (aidanheerdegen) [inline payu/sync.py:48]: I don't know if it's better/faster, but you could sort the directories and work backwards from the end (most recent/highest number) until you get a match that is `<= current_run`.

Sort touches all the elements, but it is fast. This way the `replace` and `int` comparison is done on far fewer elements. 

Just a thought.

**Comment by Qian-HuiChen** [comment]
I see. It should be fine to sync all `payu_jobs` at the current state. `restart` and `output` are the focus here.

File: payu/sync.py (hunk 1 above)
Comment (Qian-HuiChen) [reply payu/sync.py:48]: Sounds good. Implemented in [b304178](https://github.com/payu-org/payu/pull/740/commits/b304178b2b25bf4986f239317c7a78a1ddc628d5). I tried to test the time needed in each case, but both of them returns 0.00s 😆

File: test/test_sync.py [hunk 2]
```diff
@@ -83,6 +86,32 @@ def assert_expected_archive_paths(source_paths,
     assert protected_dirs == expected_protected_dirs
 
 
+def test_filter_previous_runs(monkeypatch):
+    """Test filter_previous_runs pick up runs <= the current run."""
+
+    # Set current run to 3
+    monkeypatch.setenv("PAYU_CURRENT_RUN", "3")
+
+    all_dirs = ['output001', 'output002', 'output003', 'output004', 'output005']
```
Comment (aidanheerdegen) [inline test/test_sync.py]: We [need to support the case for run numbers > 999](https://github.com/payu-org/payu/issues/232). Can you include examples with 4 digits on the end and unordered to make sure this still sorts correctly?

e.g.
```suggestion
    all_dirs = ['output1001', 'output997', 'output999', 'output1002', 'output998']
```

Which would require changing `PAYU_CURRENT_RUN` above.

File: test/test_sync.py [hunk 3]
```diff
@@ -83,6 +86,32 @@ def assert_expected_archive_paths(source_paths,
...
+    
+    assert result == expected
+
+
+def test_filter_previous_runs_no_current_run(monkeypatch):
+    """Test filter_previous_runs returns all dirs when PAYU_CURRENT_RUN is not set."""
+    
+    monkeypatch.delenv("PAYU_CURRENT_RUN", raising=False)
+
+    all_dirs = ['output001', 'output002', 'output003', 'output004', 'output005']
```
Comment (aidanheerdegen) [inline test/test_sync.py]: If changed above might make sense to do the same here.

File: payu/sync.py (hunk 1 above)
Comment (aidanheerdegen) [reply payu/sync.py:48]: You can use [`timeit`](https://docs.python.org/3.13/library/timeit.html) to get reliable stats on relatively fast functions. Also it would be very dependent on the number of elements.

**Review by aidanheerdegen — COMMENTED** [review]
Just a suggestion about testing.

File: payu/sync.py [hunk 4]
```diff
@@ -42,6 +43,30 @@ def __init__(self, path, protected=False, is_log_file=False):
...
+
+    # Sort the dicectories from highest to lowest
+    sorted_dir = sorted(all_dir)[::-1]
+    for i in range(len(sorted_dir)):
+        directory = sorted_dir[i]
+        suffix = directory.replace(prefix, '') if prefix else directory
+
+        # Only include directories that are less than or equal to the current run
+        if int(suffix) <= int(current_run):
+            return sorted_dir[i:][::-1]
```
Comment (jo-basevi) [inline payu/sync.py]: So `list_archive_dirs` already sorts the directories based on the numbers:

https://github.com/payu-org/payu/blob/27624d9d3ad07b04099068c1f40590a6d639e771/payu/fsops.py#L242-L257

A very optional suggestion (feel free to ignore this!), is to loop backwards rather than using list slices to reverse order. E.g. 

```python
for i in range(len(all_dir) - 1, -1, -1):
    directory = all_dir[i]
    suffix = directory.replace(prefix, '') if prefix else directory
    if int(suffix) <= int(current_run):
        return all_dir[:i + 1]
```

I don't think prefix ever would be None - so I would make it a required argument. 

Could use `int(d.removeprefix(prefix)` too. Today I learnt that the `list_archive_dirs` should use `removeprefix` rather than `lstrip` as `lstrip` uses removes any letters matching the characters in the string.. E.g. 

```
>>> x = "outputput111"
>>> x.removeprefix("output")
'put111'
>>> x = "outputput111"
>>> x.lstrip("output")
'111'
```

**Review by jo-basevi — COMMENTED** [review]
Thanks for adding this fix in! Changes look good to me

File: payu/sync.py (hunk 4 above)
Comment (Qian-HuiChen) [reply payu/sync.py]: Thanks for pointing out `removeprefix`! It looks very convenient. Implemented in the latest commit. Also change `list_archive_dir` with `removeprefix` and add a unit test for it.

File: test/test_sync.py (hunk 2 above)
Comment (Qian-HuiChen) [reply test/test_sync.py]: Thanks for pointing this out! It actually finds a bug:
```
sorted_dir = sorted(all_dir)
```
thinks ['output1001', 'output1002', 'output997', 'output998', 'output999'] is sorted.
I updated this line to
```
sorted_dir = sorted(all_dir, key=lambda d: int(d.removeprefix(prefix)))
```
then it knows to sort it out as ['output997', 'output998', 'output999', 'output1001', 'output1002']

File: test/test_sync.py (hunk 3 above)
Comment (Qian-HuiChen) [reply test/test_sync.py]: Implemented in the commit [a78b669](https://github.com/payu-org/payu/pull/740/commits/a78b669ac0f4e10b1ce302498d0431ccecb1d1ea).

File: payu/sync.py (hunk 4 above)
Comment (jo-basevi) [reply payu/sync.py]: There's a unit test for `test_list_archive_dirs` in test_payu.py so might be worth removing that one and keep the new one on test_fsops.py - but maybe check it wasn't testing any other edge cases?

Comment (jo-basevi) [reply payu/sync.py]: Sorry this is code I wrote a while ago but `list_archive_dirs` -> `list_sorted_archive_dirs` is probably a more appropriate name since the sorted is essential to how its used in other code in payu..

Comment (Qian-HuiChen) [reply payu/sync.py]: Oh easy, I will move the `test_list_archive_dirs` from `test_payu.py` to `test_fsops.py`. I will add a few dirs into the test for edge cases. 
I will also change the name to `list_sorted_archive_dirs`.

See commit [9d73ddb](https://github.com/payu-org/payu/pull/740/commits/9d73ddb0825b1f13be79f230446b82ed3f6f5df4)

Comment (jo-basevi) [reply payu/sync.py]: Thanks!

File: test/test_sync.py (hunk 2 above)
Comment (Qian-HuiChen) [reply test/test_sync.py]: Because @jo-basevi pointed it out that `list_archive_dir` already sorts out the directories, this should not be a problem for `filter_previous_runs`.  In the `test_filter_previous_runs`, I only put in sorted dirs for testing. The sorting process is tested in `test_list_sorted_archive_dirs` separately.

**Review by aidanheerdegen — APPROVED** [review]
Looks good. Thanks!

File: payu/sync.py [hunk 5]
```diff
@@ -42,6 +43,27 @@ def __init__(self, path, protected=False, is_log_file=False):
...
 
+def filter_previous_runs(all_dir, prefix):
+    """Given a list of directories of all runs (e.g., ['output001', 'output002']), 
+    filter to only include dir that is less than or equal to current run."""
+    if all_dir == []:
+        return []
+
+    current_run = os.environ.get('PAYU_CURRENT_RUN')
+    if current_run is None:
+        warnings.warn("PAYU_CURRENT_RUN environment variable not set. "
```
Comment (jo-basevi) [inline payu/sync.py:54]: This warning is not being displayed for me.. I've verified that that this code is reached in my test so I don't think it's an issue with this PR. 

Are warnings being displayed for you?

**Review by jo-basevi — APPROVED** [review]
Thanks @Qian-HuiChen! The changes in this PR look good to me

File: payu/sync.py (hunk 5 above)
Comment (jo-basevi) [reply payu/sync.py:54]: Ok I've confirmed that this issue is unrelated to this PR (see comment: https://github.com/payu-org/payu/pull/713#discussion_r3285098967) so shouldn't block this PR from being merged.
